src/RealtimeEffectPanel.cpp

```diff
@@ -0,0 +1,796 @@
+
+         auto sizer = std::make_unique<wxBoxSizer>(wxHORIZONTAL);
+
+         //On/off button
+         auto enableButton = safenew ThemedButtonWrapper<wxToggleButton>(this, wxID_ANY, wxEmptyString, wxDefaultPosition, wxDefaultSize, wxNO_BORDER);
+         enableButton->SetBitmapLabelIndex(bmpEffectOff);
+         enableButton->SetBitmapPressedIndex(bmpEffectOn);
+         enableButton->SetBackgroundColorIndex(clrEffectListItemBackground);
+         enableButton->SetMinSize(enableButton->GetBitmapLabel().GetSize());
+         enableButton->Bind(wxEVT_TOGGLEBUTTON, &RealtimeEffectControl::OnEnableButtonClicked, this);
+         mEnableButton = enableButton;
+
+         //Central button with effect name
+         mChangeButton = safenew ThemedButtonWrapper<wxButton>(this, wxID_ANY);
+         mChangeButton->SetMinSize(mEnableButton->GetMinSize());
+         mChangeButton->Bind(wxEVT_BUTTON, &RealtimeEffectControl::OnChangeButtonClicked, this);
+
+         //Show effect settings
+         auto optionsButton = safenew ThemedButtonWrapper<wxButton>(this, wxID_ANY, wxT(""), wxDefaultPosition, wxDefaultSize, wxNO_BORDER);
+         optionsButton->SetBitmapLabelIndex(bmpEffectSettingsNormal);
+         optionsButton->SetBitmapPressedIndex(bmpEffectSettingsDown);
+         optionsButton->SetBitmapCurrentIndex(bmpEffectSettingsHover);
+         optionsButton->SetBackgroundColorIndex(clrEffectListItemBackground);
+         optionsButton->SetMinSize(optionsButton->GetBitmapLabel().GetSize());
+         optionsButton->Bind(wxEVT_BUTTON, &RealtimeEffectControl::OnOptionsClicked, this);
+         
+         auto dragArea = safenew wxStaticBitmap(this, wxID_ANY, theTheme.Bitmap(bmpDragArea));
+         dragArea->Disable();
+         sizer->Add(
+            dragArea,
+            wxSizerFlags().Center().Border(wxLEFT, 5)
+         );
+         sizer->Add(
+            mEnableButton,
+            wxSizerFlags().Center().Border(wxLEFT, 5)
+         );
+         sizer->Add(
+            mChangeButton,
+            wxSizerFlags().Center().Border(wxLEFT, 5).Proportion(1)
+         );
+         sizer->Add(
+            optionsButton,
+            wxSizerFlags().Center().Border(wxLEFT | wxRIGHT, 5)
+         );
+
+         SetSizer(sizer.release());
+      }
+
+      void SetEffect(AudacityProject& project, const std::shared_ptr<Track>& track, RealtimeEffectState* state)
+      {
+         mProject = &project;
+         mTrack = track;
+         mEffectState = state;
+         auto desc = PluginManager::Get().GetPlugin(mEffectState->GetID());
+
+         mChangeButton->SetTranslatableLabel(desc->GetSymbol().Msgid());
+      }
+
+      void RemoveFromList()
+      {
+         if(mEffectState == nullptr)
+            return;
+         
+         auto effectName = mEffectState->GetEffect()->GetName();
+         AudioIO::Get()->RemoveState(*mProject, &*mTrack, *mEffectState);
+         ProjectHistory::Get(*mProject).PushState(
```

Comment:
I don't understand why you changed mProject to a `wxWeakRef` but still assume it is non-null anyway.

Is there any reason to think the project will ever be destroyed before the associated effects window?  Did your experiments suggest that is so?

See `ProjectManager::OnCloseWindow` where there is some care about the sequence of destruction of things.  But note the ProjectWindow is destroyed there (which should destroy its descendants), just before the last line which destroys the Project object.

So I think there is no need at all for wxWeakRef for safety.

Reply:
>I don't understand why you changed mProject to a wxWeakRef but still assume it is non-null anyway.

Right, it does not make sense

>See ProjectManager::OnCloseWindow where there is some care about the sequence of destruction of things.

`wxWindow::Destroy` does not destroy window immediately, but I'm not sure about exact order of things happening after that.

Reply:
"wxWidgets delays actual deletion of the window until all events have been processed, since otherwise there is the danger that events will be sent to a non-existent window."

https://docs.wxwidgets.org/3.0/overview_windowdeletion.html

So maybe your precaution does make sense in event handlers, but it was not fully realized without adding the null checks